pkg/kube/ready.go

```diff
@@ -272,7 +272,17 @@ func (c *ReadyChecker) volumeReady(v *corev1.PersistentVolumeClaim) bool {
 }
 
 func (c *ReadyChecker) deploymentReady(rs *appsv1.ReplicaSet, dep *appsv1.Deployment) bool {
+	// If there are no desired replicas, there will be nothing to wait for.
+	if *dep.Spec.Replicas == 0 {
+		return true
+	}
+
+	// Ensure the expected available replicas are ready.
 	expectedReady := *dep.Spec.Replicas - deploymentutil.MaxUnavailable(*dep)
+	// Also ensure at least 1 replica is ready.
+	if expectedReady < 1 {
+		expectedReady = 1
+	}
```

Comment:
I would actually like to revisit this. Sorry. As I was writing tests for this is occurred to me that if someone explicitly sets their max unavailable higher than their desired replicas than Helm would be going against the request if it didn't honor that.

So, if a user says they want 2 replicas and their max unavailable is 2 than they are saying they are OK with none being available and the app being ready. While I would not suggest doing that and an end user may not want that, the issue isn't in Helm but in what they declared if they don't want that.

Why should Helm do something other than what is explicitly declared?

Reply:
Sorry @cbf123, I found a base reasoning question.

Reply:
Who determines when something is "ready"? In the Kubernetes sense, this is codified in the resources which is what Helm reads. I don't think Helm should try to be smarter than that and override what users have specified. If a user didn't specify what they really want is a different problem.

If Helm decides to do something different from what is put in the manifests by the users we will get other bugs because of this. We would need a really good reason.

If we're talking about taking PDB into account, that's interesting. How do you see that working in to this?

Reply:
I think we should follow whatever `kubectl` does. Currently `kubectl` reports that the deployment isn't ready if not at least 1 pod is ready, even with a `maxUnavailable` of 1.